### DISCUSSION TITLE
Not enought RAM... how can I execute code in another module?

### DISCUSSION THREAD

**Discussion by nastiliano** [micropython/micropython#11243] [General]
Hello everyone,

Sorry for this essential question, but I can't find exactly what I want.

I have this kind of code inside my "main.py":

import _thread
import anotherClass #(anotherClass.py)

...

def thread_1():
    while True:
        print("thread_1")
        time.sleep(1)

def thread_2():
    anotherClass.myClass() #in myClass there's infinite llop obviously

_thread.start_new_thread(thread_1 , ())
_thread.start_new_thread(thread_2 , ())

But, I have no idea how to execute this (boot.py, main.py and anotherClass.py) correctly.
It seems like the code freezes in thread_2

Somebody can help me?
Thank you very much.

**Comment by peterhinch** [comment]
What hardware are you using?

As a general comment, threading is quite demanding on a microcontroller. Where possible it is much better to use [uasyncio](http://docs.micropython.org/en/latest/library/uasyncio.html). See also [tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md). It is much more lightweight and also more highly developed.

**Reply by nastiliano** [reply]
Hello, thanks for your response.

I'm a beginner in Python (and OOP) but I'm reading a lot (books, tutorials...) about OOP, sockets, concurrence... I wanted to use your libs (uasyncio) but I don't know how to installing it in my devkit (ESP32-S2-WROOM) :)

I think my solution it would be: keep 1 thread for my WiFi comunication and the rest of the code (E/S based code) with asyncIO.

I'm reading the uasyncio manual but I'm still unknowing how to intall it in my ESP32.

Thank you very much

**Reply by bixb922** [reply]
asyncio is already installed. You just need to ```import uasyncio```

**Reply by nastiliano** [reply]
Thanks for your response.

Mmmmm... nope. In my ESP32 (with latest micropytyhon 1.19.1: https://www.micropython.org/download/esp32/).

Using Visual Studio Code (latest version) and Pymkr extension (in order to comunicate to it) occurs an exeption error indicating that asyncio module doesn't exist nor uasyncio...

Is that normal?

Thank you

**Reply by scruss** [reply]
That would be the wrong version for an ESP32-S2-WROOM. This would be closer to the right image: [ESP32-S2](https://www.micropython.org/download/GENERIC_S2/)

**Reply by nastiliano** [reply]
Thanks for your response.

Mmmm....
I've been programming for one year (more or less correctly, without lots of errors :)
May be I'm doing my program with an incorrect micropython version?
I'm using an ESP-32 development board v1 (https://grobotronics.com/esp32-development-board-devkit-v1.html?sl=en)  and ESP32 generic micropython (https://www.micropython.org/download/esp32/).

Ok!  This night (Spain time) I'll try to change the micropython version for this: https://www.micropython.org/download/GENERIC_S2/.

Thank you very much.

**Reply by nastiliano** [reply]
Hello again.

Doesn't work. I've needed to (--force) installation of "GENERIC_S2" micropython and the board has come into infinite bucle error.
So I erased the board again.
The board chip is exactly: "ESP32-WROOM-32"

Any ideas?

Thank you very much.

**Reply by scruss** [reply]
this one: https://www.micropython.org/download/esp32/

You said it was an S2 in your earlier response. It is not an S2.

Threads are a great way to run out of memory. Best not to use them on MicroPython
